# Countries of the World

By Krzysztof Satola from: [github.com/ksatola](https://github.com/ksatola).

Based on CRISP-DM (Cross Industry Process for Data Mining).

## Business Understanding

### Dataset Dictionary

- **country** - country name
- **region** - region name
- **population** - number of people within country
- **area** - area in sq. mi.
- **popdensity** - population density per sq. mi.
- **coast** - coastline (coast/area ratio)
- **netmigr** - net migration. The net migration rate is the difference between the number of immigrants (people coming into an area) and the number of emigrants (people leaving an area) throughout the year. When the number of immigrants is larger than the number of emigrants, a positive net migration rate occurs. A positive net migration rates indicates that there are more people entering than leaving an area. When more emigrate from a country, the result is a negative net migration rate, meaning that more people are leaving than entering the area. When there is an equal number of immigrants and emigrants, the net migration rate is balanced ([source](https://en.wikipedia.org/wiki/Net_migration_rate)).
- **infmortality** - infant mortality (per 1000 births)
- **gdp** - gross domestic product (GDP) in $ per capita. The value of all final goods and services produced within a nation in a given year (2013), converted at market exchange rates to current U.S. dollars, divided by the average population for the same year ([source](https://en.wikipedia.org/wiki/List_of_countries_by_GDP_(nominal)_per_capita)).
- **literacy** - literacy level in %
- **phones** - phones per 1000
- **arable** - percent of arable areas
- **crops** - percent of cropland used to grow food
- **other** - other (%) ~????????????????~
- **climate** - climate type
- **birthrate** - the birth rate (technically, births/population rate), the total number of live births per 1,000 in a population in 2013 ([source](https://en.wikipedia.org/wiki/Birth_rate)).
- **deathrate** - number of deaths in units of deaths per 1,000 individuals ([source](https://en.wikipedia.org/wiki/Mortality_rate)).
- **agriculture** - percentage of GDP sector composition ratio for agriculture economy sector ([source](https://en.wikipedia.org/wiki/List_of_countries_by_GDP_sector_composition)). Agriculture % + Industry % + Service = 100% of GDP 
- **industry** - percentage of GDP sector composition ratio for industry economy sector
- **service** - percentage of GDP sector composition ratio for service economy sector

### Objectives

In this project, I explore [Countries of the Worlds Kaggle dataset](https://www.kaggle.com/fernandol/countries-of-the-world) to answer the following questions:

1. How different regions compared to each other in terms of area, population, population density
2. 
3. What are the most significant predictors for country's GDP ($ per capita)?

Next, I write a post to ??? about my findings.

## Data Understanding

In [25]:
# Import required libraries
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

# Random state
rstate = 123

In [11]:
%load_ext version_information

In [14]:
# Document versions of used libraries
%version_information numpy, pandas, matplotlib, seaborn

Software versions
Python 3.7.1 64bit [Clang 4.0.1 (tags/RELEASE_401/final)]
IPython 7.2.0
OS Darwin 18.2.0 x86_64 i386 64bit
numpy 1.16.2
pandas 0.23.4
matplotlib 3.0.2
seaborn 0.9.0
Tue Mar 19 22:36:10 2019 CET

In [19]:
# Load data from a CSV file
df = pd.read_csv('./data/countries of the world.csv')

In [20]:
# Initial look into the dataset
df.head().T

0  \
Country                                              Afghanistan    
Region                              ASIA (EX. NEAR EAST)            
Population                                               31056997   
Area (sq. mi.)                                             647500   
Pop. Density (per sq. mi.)                                   48,0   
Coastline (coast/area ratio)                                 0,00   
Net migration                                               23,06   
Infant mortality (per 1000 births)                         163,07   
GDP ($ per capita)                                            700   
Literacy (%)                                                 36,0   
Phones (per 1000)                                             3,2   
Arable (%)                                                  12,13   
Crops (%)                                                    0,22   
Other (%)                                                   87,65   
Climate                                                         1   
Birthrate                                                    46,6   
Deathrate                                                   20,34   
Agriculture                                                  0,38   
Industry                                                     0,24   
Service                                                      0,38   

                                                                      1  \
Country                                                        Albania    
Region                              EASTERN EUROPE                        
Population                                                      3581655   
Area (sq. mi.)                                                    28748   
Pop. Density (per sq. mi.)                                        124,6   
Coastline (coast/area ratio)                                       1,26   
Net migration                                                     -4,93   
Infant mortality (per 1000 births)                                21,52   
GDP ($ per capita)                                                 4500   
Literacy (%)                                                       86,5   
Phones (per 1000)                                                  71,2   
Arable (%)                                                        21,09   
Crops (%)                                                          4,42   
Other (%)                                                         74,49   
Climate                                                               3   
Birthrate                                                         15,11   
Deathrate                                                          5,22   
Agriculture                                                       0,232   
Industry                                                          0,188   
Service                                                           0,579   

                                                                      2  \
Country                                                        Algeria    
Region                              NORTHERN AFRICA                       
Population                                                     32930091   
Area (sq. mi.)                                                  2381740   
Pop. Density (per sq. mi.)                                         13,8   
Coastline (coast/area ratio)                                       0,04   
Net migration                                                     -0,39   
Infant mortality (per 1000 births)                                   31   
GDP ($ per capita)                                                 6000   
Literacy (%)                                                       70,0   
Phones (per 1000)                                                  78,1   
Arable (%)                                                         3,22   
Crops (%)                                                      

In [22]:
df.tail().T

222  \
Country                                                      West Bank    
Region                              NEAR EAST                             
Population                                                      2460492   
Area (sq. mi.)                                                     5860   
Pop. Density (per sq. mi.)                                        419,9   
Coastline (coast/area ratio)                                       0,00   
Net migration                                                      2,98   
Infant mortality (per 1000 births)                                19,62   
GDP ($ per capita)                                                  800   
Literacy (%)                                                        NaN   
Phones (per 1000)                                                 145,2   
Arable (%)                                                         16,9   
Crops (%)                                                         18,97   
Other (%)                                                         64,13   
Climate                                                               3   
Birthrate                                                         31,67   
Deathrate                                                          3,92   
Agriculture                                                        0,09   
Industry                                                           0,28   
Service                                                            0,63   

                                                                    223  \
Country                                                 Western Sahara    
Region                              NORTHERN AFRICA                       
Population                                                       273008   
Area (sq. mi.)                                                   266000   
Pop. Density (per sq. mi.)                                          1,0   
Coastline (coast/area ratio)                                       0,42   
Net migration                                                       NaN   
Infant mortality (per 1000 births)                                  NaN   
GDP ($ per capita)                                                  NaN   
Literacy (%)                                                        NaN   
Phones (per 1000)                                                   NaN   
Arable (%)                                                         0,02   
Crops (%)                                                             0   
Other (%)                                                         99,98   
Climate                                                               1   
Birthrate                                                           NaN   
Deathrate                                                           NaN   
Agriculture                                                         NaN   
Industry                                                            NaN   
Service                                                             0,4   

                                                                    224  \
Country                                                          Yemen    
Region                              NEAR EAST                             
Population                                                     21456188   
Area (sq. mi.)                                                   527970   
Pop. Density (per sq. mi.)                                         40,6   
Coastline (coast/area ratio)                                       0,36   
Net migration                                                         0   
Infant mortality (per 1000 births)                                 61,5   
GDP ($ per capita)                                                  800   
Literacy (%)                                                       50,2   
Phones (per 1000)                                                  37,2   
Arable (%)      

In [26]:
df.sample(5, random_state=rstate).T

125  \
Country                                                         Malawi    
Region                              SUB-SAHARAN AFRICA                    
Population                                                     13013926   
Area (sq. mi.)                                                   118480   
Pop. Density (per sq. mi.)                                        109,8   
Coastline (coast/area ratio)                                       0,00   
Net migration                                                         0   
Infant mortality (per 1000 births)                               103,32   
GDP ($ per capita)                                                  600   
Literacy (%)                                                       62,7   
Phones (per 1000)                                                   7,9   
Arable (%)                                                        23,38   
Crops (%)                                                          1,49   
Other (%)                                                         75,13   
Climate                                                               2   
Birthrate                                                         43,13   
Deathrate                                                         19,33   
Agriculture                                                       0,342   
Industry                                                          0,158   
Service                                                           0,499   

                                                              122  \
Country                                                    Macau    
Region                              ASIA (EX. NEAR EAST)            
Population                                                 453125   
Area (sq. mi.)                                                 28   
Pop. Density (per sq. mi.)                                16183,0   
Coastline (coast/area ratio)                               146,43   
Net migration                                                4,86   
Infant mortality (per 1000 births)                           4,39   
GDP ($ per capita)                                          19400   
Literacy (%)                                                 94,5   
Phones (per 1000)                                           384,9   
Arable (%)                                                      0   
Crops (%)                                                       0   
Other (%)                                                     100   
Climate                                                         2   
Birthrate                                                    8,48   
Deathrate                                                    4,47   
Agriculture                                                 0,001   
Industry                                                    0,072   
Service                                                     0,927   

                                                              156  \
Country                                                 Pakistan    
Region                              ASIA (EX. NEAR EAST)            
Population                                              165803560   
Area (sq. mi.)                                             803940   
Pop. Density (per sq. mi.)                                  206,2   
Coastline (coast/area ratio)                                 0,13   
Net migration                                               -2,77   
Infant mortality (per 1000 births)                          72,44   
GDP ($ per capita)                                           2100   
Literacy (%)                                                 45,7   
Phones (per 1000)                                            31,8   
Arable (%)                                                  27,87   
Crops (%)                                                    0,87   
Other (%)                                                   71,26   
Climate

In [4]:
# Dataset size
df.shape

(227, 20)

In [8]:
# Variables
df.columns

Index(['Country', 'Region', 'Population', 'Area (sq. mi.)',
       'Pop. Density (per sq. mi.)', 'Coastline (coast/area ratio)',
       'Net migration', 'Infant mortality (per 1000 births)',
       'GDP ($ per capita)', 'Literacy (%)', 'Phones (per 1000)', 'Arable (%)',
       'Crops (%)', 'Other (%)', 'Climate', 'Birthrate', 'Deathrate',
       'Agriculture', 'Industry', 'Service'],
      dtype='object')

In [29]:
# Naming convention, simplify column names and build a dataset dictionary (see above)
df.rename(columns={"Country":"country", 
                  "Region":"region", 
                  "Population":"population", 
                  "Area (sq. mi.)":"area", 
                  "Pop. Density (per sq. mi.)":"popdensity", 
                  "Coastline (coast/area ratio)":"coast", 
                  "Net migration":"netmigr", 
                  "Infant mortality (per 1000 births)":"infmortality", 
                  "GDP ($ per capita)":"gdp", 
                  "Literacy (%)":"literacy", 
                  "Phones (per 1000)":"phones", 
                  "Arable (%)":"arable", 
                  "Crops (%)":"crops", 
                  "Other (%)":"other", 
                  "Climate":"climate", 
                  "Birthrate":"birthrate", 
                  "Deathrate":"deathrate", 
                  "Agriculture":"agriculture", 
                  "Industry":"industry", 
                  "Service":"service"}, inplace=True)

In [30]:
df.columns

Index(['country', 'region', 'population', 'area', 'popdensity', 'coast',
       'netmigr', 'infmortality', 'gdp', 'literacy', 'phones', 'arable',
       'crops', 'other', 'climate', 'birthrate', 'deathrate', 'agriculture',
       'industry', 'service'],
      dtype='object')

In [39]:
# Examplary country data
df.iloc[163]

country                                     Poland 
region          EASTERN EUROPE                     
population                                 38536869
area                                         312685
popdensity                                    123,3
coast                                          0,16
netmigr                                       -0,49
infmortality                                   8,51
gdp                                           11100
literacy                                       99,8
phones                                        306,3
arable                                        45,91
crops                                          1,12
other                                         52,97
climate                                           3
birthrate                                      9,85
deathrate                                      9,89
agriculture                                    0,05
industry                                      0,311
service     

In [44]:
# What are the dataset column data types?
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 227 entries, 0 to 226
Data columns (total 20 columns):
country         227 non-null object
region          227 non-null object
population      227 non-null int64
area            227 non-null int64
popdensity      227 non-null object
coast           227 non-null object
netmigr         224 non-null object
infmortality    224 non-null object
gdp             226 non-null float64
literacy        209 non-null object
phones          223 non-null object
arable          225 non-null object
crops           225 non-null object
other           225 non-null object
climate         205 non-null object
birthrate       224 non-null object
deathrate       223 non-null object
agriculture     212 non-null object
industry        211 non-null object
service         212 non-null object
dtypes: float64(1), int64(2), object(17)
memory usage: 35.5+ KB


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 227 entries, 0 to 226
Data columns (total 20 columns):
country         227 non-null object
region          227 non-null object
population      227 non-null int64
area            227 non-null int64
popdensity      227 non-null object
coast           227 non-null object
netmigr         224 non-null object
infmortality    224 non-null object
gdp             226 non-null float64
literacy        209 non-null object
phones          223 non-null object
arable          225 non-null object
crops           225 non-null object
other           225 non-null object
climate         205 non-null object
birthrate       224 non-null object
deathrate       223 non-null object
agriculture     212 non-null object
industry        211 non-null object
service         212 non-null object
dtypes: float64(1), int64(2), object(17)
memory usage: 35.5+ KB


### Notes

1. The dataset has 20 variables and 227 observations (one per each country).
2. The dataset column names were standardized and their meaning described in the dataset dictionary.
3. The dataset column types were corrected.
4. The dataset columns with missing values are

## Prepare Data

## Data Modeling??

## Evaluate the Result

## Deploy